src/audit/routes/maintain.py

```diff
@@ -0,0 +1,139 @@
+        > for k, v in row.items():
+        E  AttributeError: 'NoneType' object has no attribute 'items'
+    In later versions, it results in a `NoSuchRowError` exception
+    (https://github.com/python-gino/gino/blob/v1.1.0b2/src/gino/crud.py#L858).
+    However, at the time of writing this update has only been pre-released.
+
+    According to this doc
+    https://wiki.postgresql.org/wiki/INSERT_RETURNING_vs_Partitioning, we could
+    implement a workaround. But since we do not currently need GINO to return
+    the inserted statement, and the row does get inserted before the exception
+    is raised, we just catch and ignore it.
+
+    TODO Once a newer version of GINO is released, upgrade and catch
+    `NoSuchRowError` instead of `AttributeError`.
+    """
+    try:
+        await CATEGORY_TO_MODEL_CLASS[category].create(**data)
+    except AttributeError:
+        pass
+
+
+def handle_timestamp(data):
+    """
+    If the timestamp is omitted from the request body, we use the current date
+    and time. In most cases, services should NOT provide a timestamp when
+    creating audit logs. The timestamp is only accepted in log creation
+    requests to allow populating the audit database with historical data, for
+    example by parsing historical logs from before the Audit Service was
+    deployed to a Data Commons.
+    """
+    # "timestamp" key always exists because it's defined in `CreateLogInput`
+    if data["timestamp"]:
+        # we take a timestamp as input, but store a datetime in the database
+        data["timestamp"] = datetime.fromtimestamp(data["timestamp"])
+    else:
+        # timestamp=now is automatically added to rows without timestamp,
+        # but we have to remove the key from the data dict
+        del data["timestamp"]
+
+
+@router.post("/log/presigned_url", status_code=HTTP_201_CREATED)
+async def create_presigned_url_log(
+    body: CreatePresignedUrlLogInput,
+    background_tasks: BackgroundTasks,
+    auth=Depends(Auth),
+) -> dict:
+    """
+    Create a new `presigned_url` audit log.
+
+    This endpoint does not include any authorization checks, but it is not
+    exposed and is only meant for internal use.
+
+    If the timestamp is omitted from the request body, the current date and
+    time will be used.
+
+    The response is returned _before_ inserting the new audit log in the
+    database, so that POSTing audit logs does not impact the performance of
+    the caller and audit-service failures are not visible to users.
+    """
+    data = body.dict()
+    logger.debug(f"Creating `presigned_url` audit log. Received body: {data}")
+
+    allowed_actions = ["download", "upload"]
+    if data["action"] not in allowed_actions:
+        raise HTTPException(
+            HTTP_400_BAD_REQUEST,
+            f"Action '{data['action']}' is not allowed ({allowed_actions})",
+        )
+
+    handle_timestamp(data)
+    background_tasks.add_task(insert_row, "presigned_url", data)
```

Comment:
what happens if adding fails?

Reply:
and same for every other place we're using this.

Reply:
Yeah fair point, this will probably be refactored as part of the ticket to improve monitoring - see [this doc](https://github.com/uc-cdis/audit-service/blob/ed94eae115854878473f5e1abcea3acef1225f2c/docs/explanation/async_post.md)

Reply:
Oh got it.